rfd/0080-tls-ping.md

```diff
@@ -0,0 +1,415 @@
+supported by Teleport provide such features.
+
+## Details
+
+To avoid having idle connections, we need to send at least one byte before the
+idle timeout period elapses. For protocols that that doesn't offer a mechanism
+for transmitting data in pre-defined intervals (avoiding the connection from
+becoming idle), it is necessary to wrap them into a different protocol.
+
+### Ping protocol
+
+A light protocol enables the connections to send ping packets filtered out on
+the receiver. These packets can be sent periodically, preventing the connection
+from becoming idle.
+
+The protocol packet is defined by the following:
+```
+| length | data           |
+| uint32 | `length` bytes |
+```
+
+The `length` field is an `uint32` encoded in network order (big-endian).
+
+#### Ping message
+
+The ping packet consists of a message where `length = 0`, and there is no `data`
+present. These messages are not visible to the reader since they have no
+content.
+
+#### Data message
+
+When sending data messages, its content size must be encoded and sent as the
+package length. Nothing is sent if the data length is equal to `0` since it is
+used to identify ping messages.
+
+#### PoC implementation
+
+<details>
+<summary>View code</summary>
+
+```go
+// pingConn wraps a net.Conn and add ping capabilities to it, including the
+// `WritePing` function and `Read` (which excludes ping packets).
+//
+// When using this connection, the packets written will contain an initial data:
+// the packet size. When reading, this information is taken into account, but it
+// is not returned to the caller.
+//
+// Ping messages have a packet size of zero and are produced only when
+// `WritePing` is called. On `Read`, any Ping packet is discarded.
+type pingConn struct {
+	net.Conn
+
+	muRead  sync.Mutex
+	muWrite sync.Mutex
+
+	// bytesRead number of bytes already read from the current packet.
+	bytesRead int
+	// currentSize size of bytes of the current packet.
+	currentSize uint32
+}
+
+func (c *pingConn) Read(p []byte) (int, error) {
+	c.muRead.Lock()
+	defer c.muRead.Unlock()
+
+	err := c.discardPingReads()
+	if err != nil {
+		return 0, err
+	}
+
+	// Check if the current size is larger than the provided buffer.
+	readSize := c.currentSize
+	if c.currentSize > uint32(len(p)) {
+		readSize = int32(len(p))
+	}
+
+	n, err := c.Conn.Read(p[:readSize])
+	c.bytesRead += n
+
+	// Check if it has read everything.
+	if uint32(c.bytesRead) >= c.currentSize {
+		c.bytesRead = 0
+		c.currentSize = 0
+	}
+
+	return n, err
+}
+
+// WritePing writes the ping packet to the connection.
+func (c *pingConn) WritePing() error {
+	c.muWrite.Lock()
+	defer c.muWrite.Unlock()
+
+	return binary.Write(c.Conn, binary.BigEndian, uint32(0))
+}
+
+// discardPingReads reads from the wrapped net.Conn until it encounters a
+// non-ping packet.
+func (c *pingConn) discardPingReads() error {
+	if c.bytesRead > 0 {
+		return nil
+	}
+
+	for c.currentSize == 0 {
```

Comment:
Not sure I fully understand this logic. Shouldn't this be "if" instead of "for"? If the first 4 bytes of each packet are always length.